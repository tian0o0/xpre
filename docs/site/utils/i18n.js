// component mixin
import { get } from '../../../src/utils';
import { camelize } from '../../../src/utils/format/string';

export default {
  computed: {
    $t() {
      const { name } = this.$options;
      const prefix = name ? camelize(name) + '.' : '';
      const messages = this.$vantMessages[this.$vantLang];

      return (path, ...args) => {
        const message = get(messages, prefix + path) || get(messages, path);
        return typeof message === 'function' ? message(...args) : message;
      };
    }
  }
};
