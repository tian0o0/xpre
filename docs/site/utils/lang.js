import Locale from '../../../src/components/locale';
import zhCN from '../../../src/components/locale/lang/zh-CN';
import enUS from '../../../src/components/locale/lang/en-US';

const langMap = {
  'en-US': {
    title: 'Xpre - Exploring Endless Possibilities',
    messages: enUS
  },
  'zh-CN': {
    title: 'Xpre | 一个差不多的组件库',
    messages: zhCN
  }
};

let currentLang = '';

function getDefaultLang() {
  const langs = Object.keys(langMap);
  const { hash } = location;

  for (let i = 0; i < langs.length; i++) {
    if (hash.indexOf(langs[i]) !== -1) {
      return langs[i];
    }
  }

  const userLang = localStorage.getItem('VANT_LANGUAGE') || navigator.language || 'en-US';
  return userLang.indexOf('zh-') !== -1 ? 'zh-CN' : 'en-US';
}

export function setLang(lang) {
  if (currentLang === lang) {
    return;
  }

  currentLang = lang;
  if (window.localStorage) {
    localStorage.setItem('VANT_LANGUAGE', lang);
  }
  Locale.use(lang, langMap[lang].messages);
  document.title = langMap[lang].title;
}

setLang(getDefaultLang());
