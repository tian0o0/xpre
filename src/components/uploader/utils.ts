export function toArray<T>(item: T | T[]): T[] {
  if (Array.isArray(item)) {
    return item;
  }

  return [item];
}

export function readFile(file: File, resultType: string) {
  return new Promise(resolve => {
    const reader = new FileReader();

    reader.onload = event => {
      resolve((event.target as FileReader).result);
    };

    if (resultType === 'dataUrl') {
      reader.readAsDataURL(file);
    } else if (resultType === 'text') {
      reader.readAsText(file);
    }
  });
}

export function isOversize(files: File | File[], maxSize: number): boolean {
  return toArray(files).some(file => file.size > maxSize);
}

export type FileListItem = {
  url?: string;
  file?: File;
  content?: string; // dataUrl
};

const IMAGE_EXT = ['jpeg', 'jpg', 'gif', 'png', 'svg'];

export function isImageUrl(url: string): boolean {
  return IMAGE_EXT.some(ext => url.indexOf(`.${ext}`) !== -1);
}

export function isImageFile(item: FileListItem): boolean {
  if (item.file && item.file.type) {
    return item.file.type.indexOf('image') === 0;
  }

  if (item.url) {
    return isImageUrl(item.url);
  }

  if (item.content) {
    return item.content.indexOf('data:image') === 0;
  }

  return false;
}
