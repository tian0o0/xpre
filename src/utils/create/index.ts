import { createBEM, Bem } from './bem';
import { createComponent, Component } from './component';
import { createI18N, Translate } from './i18n';

type CreateNamespaceReturn = [
  Component,
  Bem,
  Translate
];

export function createNamespace(name: string): CreateNamespaceReturn {
  name = 'x-' + name;
  return [createComponent(name), createBEM(name), createI18N(name)];
}
