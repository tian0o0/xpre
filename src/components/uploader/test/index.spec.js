import Uploader from '..';
import { mount, later } from '../../../../test/utils';

window.File = function () {
  this.size = 10000;
};

const mockFileDataUrl = 'data:image/test';
const mockFile = new File([], 'test.jpg');
const file = { target: { files: [mockFile] } };
const multiFile = { target: { files: [mockFile, mockFile] } };

window.FileReader = function () {
  this.readAsText = function () {
    this.onload &&
      this.onload({
        target: {
          result: mockFileDataUrl
        }
      });
  };
  this.readAsDataURL = this.readAsText;
};

test('disabled', () => {
  const afterRead = jest.fn();
  const wrapper = mount(Uploader, {
    propsData: {
      disabled: true,
      afterRead
    }
  });

  wrapper.vm.onChange(file);
  expect(afterRead).toHaveBeenCalledTimes(0);
});

it('read text', done => {
  const wrapper = mount(Uploader, {
    propsData: {
      resultType: 'text',
      afterRead: readFile => {
        expect(readFile.content).toEqual(mockFileDataUrl);
        done();
      }
    }
  });

  wrapper.vm.onChange(file);
});

// it('set input name', done => {
//   const wrapper = mount(Uploader, {
//     propsData: {
//       name: 'uploader',
//       beforeRead: (readFile, detail) => {
//         expect(detail.name).toEqual('uploader');
//         return true;
//       },
//       afterRead: (readFile, detail) => {
//         expect(detail.name).toEqual('uploader');
//         done();
//       }
//     }
//   });

//   wrapper.vm.onChange(file);
// });

it('unknown resultType', () => {
  const afterRead = jest.fn();
  const wrapper = mount(Uploader, {
    propsData: {
      resultType: 'xxxx',
      afterRead
    }
  });
  wrapper.vm.onChange(file);
  expect(afterRead).toHaveBeenCalledTimes(0);
});

it('before read return false', () => {
  const afterRead = jest.fn();
  const wrapper = mount(Uploader, {
    propsData: {
      beforeRead: () => false,
      afterRead
    }
  });

  const input = wrapper.find('input');

  wrapper.vm.onChange(file);
  expect(afterRead).toHaveBeenCalledTimes(0);
  expect(input.element.value).toEqual('');
});

it('before read return promise and resolve', async () => {
  const afterRead = jest.fn();
  const wrapper = mount(Uploader, {
    propsData: {
      beforeRead: () => new Promise(resolve => {
        resolve();
      }),
      afterRead
    }
  });

  wrapper.vm.onChange(file);

  await later();
  expect(afterRead).toHaveBeenCalledTimes(1);
});

it('before read return promise and reject', async () => {
  const afterRead = jest.fn();
  const wrapper = mount(Uploader, {
    propsData: {
      beforeRead: () => new Promise((resolve, reject) => {
        reject();
      }),
      afterRead
    }
  });

  const input = wrapper.find('input');

  await later();
  wrapper.vm.onChange(file);
  expect(afterRead).toHaveBeenCalledTimes(0);
  expect(input.element.value).toEqual('');
});

test('file size overlimit', async () => {
  const wrapper = mount(Uploader, {
    propsData: {
      maxSize: 1
    }
  });
  wrapper.vm.onChange(file);
  wrapper.vm.onChange(multiFile);

  await later();
  expect(wrapper.emitted('oversize')[0]).toBeTruthy();
  expect(wrapper.emitted('oversize')[1]).toBeTruthy();

  wrapper.vm.maxSize = 100000;
  wrapper.vm.onChange(multiFile);

  await later();
  expect(wrapper.emitted('oversize')[2]).toBeFalsy();
});

it('render upload-text', () => {
  const wrapper = mount(Uploader, {
    propsData: {
      uploadText: 'Text'
    }
  });

  expect(wrapper).toMatchSnapshot();
});

it('render preview image', async () => {
  const wrapper = mount(Uploader, {
    propsData: {
      fileList: [
        { url: 'https://ae01.alicdn.com/kf/H6e5e22c9a0214cf692a69097c342e981T.png' },
        { url: 'https://img.yzcdn.cn/vant/test.pdf' },
        { file: { name: 'test.pdf' } }
      ]
    },
    listeners: {
      input(fileList) {
        wrapper.setProps({ fileList });
      }
    }
  });

  wrapper.vm.onChange(file);
  await later();

  expect(wrapper).toMatchSnapshot();
});

it('disable preview image', async () => {
  const wrapper = mount(Uploader, {
    propsData: {
      fileList: [],
      previewImage: false
    },
    listeners: {
      input(fileList) {
        wrapper.setProps({ fileList });
      }
    }
  });

  wrapper.vm.onChange(file);
  await later();

  expect(wrapper).toMatchSnapshot();
});

it('max-count prop', async () => {
  const wrapper = mount(Uploader, {
    propsData: {
      fileList: [],
      maxCount: 1
    },
    listeners: {
      input(fileList) {
        wrapper.setProps({ fileList });
      }
    }
  });

  wrapper.vm.onChange(multiFile);
  await later();

  expect(wrapper).toMatchSnapshot();
});

it('preview-size prop', async () => {
  const wrapper = mount(Uploader, {
    propsData: {
      fileList: [],
      previewSize: 30
    },
    listeners: {
      input(fileList) {
        wrapper.setProps({ fileList });
      }
    }
  });

  wrapper.vm.onChange(file);
  await later();

  expect(wrapper).toMatchSnapshot();
});

it('delete preview image', async () => {
  const wrapper = mount(Uploader, {
    propsData: {
      fileList: [],
      previewSize: 30
    },
    listeners: {
      input(fileList) {
        wrapper.setProps({ fileList });
      }
    }
  });

  wrapper.vm.onChange(file);
  await later();

  wrapper.find('.x-uploader__preview-delete').trigger('click');
  expect(wrapper.vm.fileList.length).toEqual(0);

  expect(wrapper).toMatchSnapshot();
  expect(wrapper.emitted('delete')[0]).toBeTruthy();
});

it('click to preview image', async () => {
  const wrapper = mount(Uploader, {
    propsData: {
      fileList: [
        { url: 'https://ae01.alicdn.com/kf/H6e5e22c9a0214cf692a69097c342e981T.png' },
        { url: 'https://img.yzcdn.cn/vant/test.pdf' }
      ],
      previewSize: 30
    },
    listeners: {
      input(fileList) {
        wrapper.setProps({ fileList });
      }
    }
  });

  wrapper.vm.onChange(file);
  await later();

  wrapper.find('.x-image').trigger('click');

//   const imagePreviewNode = document.querySelector('.x-image-preview');
//   expect(imagePreviewNode).toMatchSnapshot();
//   imagePreviewNode.remove();
});
