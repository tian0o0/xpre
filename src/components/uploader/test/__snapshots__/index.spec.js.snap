// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delete preview image 1`] = `
<div class="x-uploader">
  <div class="x-uploader__wrapper">
    <div class="x-uploader__upload" style="width: 30px; height: 30px;"><i class="x-icon x-icon-plus x-uploader__upload-icon"></i><input type="file" accept="image/*" class="x-uploader__input"></div>
  </div>
</div>
`;

exports[`disable preview image 1`] = `
<div class="x-uploader">
  <div class="x-uploader__wrapper">
    <div class="x-uploader__upload"><i class="x-icon x-icon-plus x-uploader__upload-icon"></i><input type="file" accept="image/*" class="x-uploader__input"></div>
  </div>
</div>
`;

exports[`max-count prop 1`] = `
<div class="x-uploader">
  <div class="x-uploader__wrapper">
    <div class="x-uploader__preview">
      <div class="x-image x-uploader__preview-image"><img src="data:image/test" class="x-image__img" style="object-fit: cover;">
        <div class="x-image__loading"><i class="x-icon x-icon-photo-o" style="font-size: 22px;"></i></div>
      </div><i class="x-icon x-icon-delete x-uploader__preview-delete"></i>
    </div>
  </div>
</div>
`;

exports[`preview-size prop 1`] = `
<div class="x-uploader">
  <div class="x-uploader__wrapper">
    <div class="x-uploader__preview" style="">
      <div class="x-image x-uploader__preview-image" style="width: 30px; height: 30px;"><img src="data:image/test" class="x-image__img" style="object-fit: cover;">
        <div class="x-image__loading"><i class="x-icon x-icon-photo-o" style="font-size: 22px;"></i></div>
      </div><i class="x-icon x-icon-delete x-uploader__preview-delete"></i>
    </div>
    <div class="x-uploader__upload" style="width: 30px; height: 30px;"><i class="x-icon x-icon-plus x-uploader__upload-icon"></i><input type="file" accept="image/*" class="x-uploader__input"></div>
  </div>
</div>
`;

exports[`render preview image 1`] = `
<div class="x-uploader">
  <div class="x-uploader__wrapper">
    <div class="x-uploader__preview">
      <div class="x-image x-uploader__preview-image"><img src="https://img.yzcdn.cn/vant/cat.jpeg" class="x-image__img" style="object-fit: cover;">
        <div class="x-image__loading"><i class="x-icon x-icon-photo-o" style="font-size: 22px;"></i></div>
      </div><i class="x-icon x-icon-delete x-uploader__preview-delete"></i>
    </div>
    <div class="x-uploader__preview">
      <div class="x-uploader__file"><i class="x-icon x-icon-description x-uploader__file-icon"></i>
        <div class="x-uploader__file-name x-ellipsis">https://img.yzcdn.cn/vant/test.pdf</div>
      </div><i class="x-icon x-icon-delete x-uploader__preview-delete"></i>
    </div>
    <div class="x-uploader__preview">
      <div class="x-uploader__file"><i class="x-icon x-icon-description x-uploader__file-icon"></i>
        <div class="x-uploader__file-name x-ellipsis">test.pdf</div>
      </div><i class="x-icon x-icon-delete x-uploader__preview-delete"></i>
    </div>
    <div class="x-uploader__preview">
      <div class="x-image x-uploader__preview-image"><img src="data:image/test" class="x-image__img" style="object-fit: cover;">
        <div class="x-image__loading"><i class="x-icon x-icon-photo-o" style="font-size: 22px;"></i></div>
      </div><i class="x-icon x-icon-delete x-uploader__preview-delete"></i>
    </div>
    <div class="x-uploader__upload"><i class="x-icon x-icon-plus x-uploader__upload-icon"></i><input type="file" accept="image/*" class="x-uploader__input"></div>
  </div>
</div>
`;

exports[`render upload-text 1`] = `
<div class="x-uploader">
  <div class="x-uploader__wrapper">
    <div class="x-uploader__upload"><i class="x-icon x-icon-plus x-uploader__upload-icon"></i><span class="x-uploader__upload-text">Text</span><input type="file" accept="image/*" class="x-uploader__input"></div>
  </div>
</div>
`;
